akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowFutureFlowSpec.scala

```diff
@@ -0,0 +1,527 @@
+        }
+
+        fNotUsed.futureValue should be(NotUsed)
+        fSeq.futureValue should equal(99 +: (0 until 10))
+
+      }
+
+      "handle closed downstream when flow future is pre completed" in assertAllStagesStopped {
+        val (fSeq1, fSeq2) = src10()
+          .viaMat {
+            Flow.futureFlow {
+              Future.successful {
+                Flow[Int].alsoToMat(Sink.seq)(Keep.right)
+              }
+            }
+          }(Keep.right)
+          .mapMaterializedValue(_.flatten)
+          .take(0)
+          .toMat(Sink.seq)(Keep.both)
+          .withAttributes(attributes)
+          .run()
+
+        fSeq1.futureValue should be(empty)
+        fSeq2.futureValue should be(empty)
+
+      }
+
+      "handle closed downstream when flow future is late completed" in assertAllStagesStopped {
+        val prFlow = Promise[Flow[Int, Int, Future[collection.immutable.Seq[Int]]]]
+        val (fSeq1, fSeq2) = src10()
+          .viaMat {
+            Flow.futureFlow(prFlow.future)
+          }(Keep.right)
+          .mapMaterializedValue(_.flatten)
+          .take(0)
+          .toMat(Sink.seq)(Keep.both)
+          .withAttributes(attributes)
+          .run()
+
+        if (delayDownstreanCancellation) {
+          fSeq1.value should be(empty)
+          fSeq2.value should be(empty)
+
+          prFlow.success {
+            Flow[Int].alsoToMat(Sink.seq)(Keep.right)
+          }
+
+          fSeq1.futureValue should be(empty)
+          fSeq2.futureValue should be(empty)
+        } else {
+          fSeq1.failed.futureValue should be(a[NeverMaterializedException])
+          fSeq1.failed.futureValue.getCause should be(a[NonFailureCancellation])
+          fSeq2.futureValue should be(empty)
+        }
+      }
+
+      "handle early downstream failure when flow future is pre-completed" in assertAllStagesStopped {
+        val (fSeq1, fSeq2) = src10()
+          .viaMat {
+            Flow.futureFlow {
+              Future.successful {
+                Flow[Int].alsoToMat(Sink.seq)(Keep.right)
+              }
+            }
+          }(Keep.right)
+          .mapMaterializedValue(_.flatten)
+          .prepend(Source.failed(TE("damn!")))
+          .toMat(Sink.seq)(Keep.both)
+          .withAttributes(attributes)
+          .run()
+
+        fSeq1.failed.futureValue should equal(TE("damn!"))
+        fSeq2.failed.futureValue should equal(TE("damn!"))
+      }
+
+      "handle early downstream failure when flow future is late completed" in assertAllStagesStopped {
+        val prFlow = Promise[Flow[Int, Int, Future[collection.immutable.Seq[Int]]]]
+        val (fSeq1, fSeq2) = src10()
+          .viaMat {
+            Flow.futureFlow(prFlow.future)
+          }(Keep.right)
+          .mapMaterializedValue(_.flatten)
+          .prepend(Source.failed(TE("damn!")))
+          .toMat(Sink.seq)(Keep.both)
+          .withAttributes(attributes)
+          .run()
+
+        if (delayDownstreanCancellation) {
```

Comment:
It's a bit unusual to structure tests like this but maybe it's ok for now.

Reply:
this was the best way I could think of to stay DRY as possible.
please notice most tests were **not** affected by the newly introduced attribute, nevertheless these still must be tested in both modes to catch any possible regression.

Reply:
Yep, I agree.